sdk/core/azure-core/src/main/java/com/azure/core/util/polling/PollingUtil.java

```diff
@@ -46,18 +45,27 @@ static <T> PollResponse<T> pollingLoop(PollingContext<T> pollingContext, Duratio
             PollResponse<T> pollResponse1 = pollOperation.apply(pollingContext);
             pollingContext.setLatestResponse(pollResponse1);
         };
+
         while (!intermediatePollResponse.getStatus().isComplete()) {
             long elapsedTime = System.currentTimeMillis() - startTime;
-            if (timeBound ?  elapsedTime >= timeoutInMillis : false) {
+            if (timeBound && elapsedTime >= timeoutInMillis) {
                 scheduler.shutdown();
-                return intermediatePollResponse;
+                if (intermediatePollResponse.getStatus().equals(statusToWaitFor)) {
+                    return intermediatePollResponse;
+                } else {
+                    throw LOGGER.logExceptionAsError(new RuntimeException(
+                        new TimeoutException("Polling didn't complete before the timeout period.")));
+                }
             }
-            if (statusToWaitFor != null && intermediatePollResponse.getStatus().equals(statusToWaitFor)) {
+
+            if (intermediatePollResponse.getStatus().equals(statusToWaitFor)) {
```

Comment:
We've got the same problem here as we might poll until completion if we didn't poll the service when the operation reached the expected state and had moved on from that state. However, we don't have a solution to this and we'll continue polling to reach the completion state.